File: estimate/event/estimate_panel.py
import pandas as pd
import numpy as np
import statsmodels.api as sm
from linearmodels.panel.data import PanelData
from linearmodels.panel import PanelOLS, PooledOLS, RandomEffects, compare
import matplotlib.pyplot as plt
from plot_result import plot_result
df=pd.read_csv("../../data/merged/len3_ship_DN.csv",encoding="shift-jis",index_col=0)
# sum ver.
# df=pd.read_csv("../data/merged/len3_ship_sum_DN.csv",encoding="shift-jis",index_col=0)
# normalization of year
# 2008で正規化するかで結構estimateが変わります、1ぐらい
# df["year"]=df["year"]-2008

#ここはかなり重要でセレクションバイアスみたいなので有意性が変わるらしい
df=df.loc[df["otc_dom"]>0]
df=df.loc[df["rx_dom"]>0]

# make log columns
df["log_rx_dom"]=np.log(df["rx_dom"])
df["log_otc_dom"]=np.log(df["otc_dom"])
df["log_stock_rx_lag"]=np.log(df["stock_rx_lag"]+0.001)
df["log_stock_otc_lag"]=np.log(df["stock_otc_lag"]+0.001)
df["year_trend"]=df["year"]-2008
df=df.set_index(["id","year"])
df.info()
# make this to paneldata
df=PanelData(df)
df.shape
# formula="rx_dom ~ year_trend+elapsed_m10+elapsed_m9+elapsed_m8+elapsed_m7+elapsed_m6+elapsed_m5+elapsed_m4+elapsed_m3+elapsed_m2+elapsed_0+elapsed_1+elapsed_2+elapsed_3+elapsed_4+elapsed_5+elapsed_6+elapsed_7+elapsed_8+elapsed_9+elapsed_10+elapsed_11+elapsed_12+elapsed_13+elapsed_14+elapsed_15+generic_per+TimeEffects+EntityEffects"
# # stock lag version
# # formula+="+stock_rx_lag"
# # gemeric_per
# formula+=" + generic_per"
# mod_fe=PanelOLS.from_formula(formula, data=df,drop_absorbed=True)
# res_fe=mod_fe.fit(cov_type='clustered', cluster_entity=True)
# # print(res_fe.summary.tables[1])
# # print(res_fe.summary.tables[0])
# print(res_fe)
# plot_result(res_fe,file_name="rx",title="rx",period=25,insert_index=9,time_start=-10)
# # print(x)
[0]*5
# take log
formula="log_rx_dom ~ year_trend+elapsed_m10+elapsed_m9+elapsed_m8+elapsed_m7+elapsed_m6+elapsed_m5+elapsed_m4+elapsed_m3+elapsed_m2+elapsed_0+elapsed_1+elapsed_2+elapsed_3+elapsed_4+elapsed_5+elapsed_6+elapsed_7+elapsed_8+elapsed_9+elapsed_10+elapsed_11+elapsed_12+elapsed_13+elapsed_14+elapsed_15+TimeEffects+EntityEffects"
# stock lag version
formula+="+log_stock_rx_lag"

# generic_per
formula+=" + generic_per"

# generic_share_r
formula+=" + generic_share_r"

# generic_share_q
formula+=" + generic_share_q"

mod_fe=PanelOLS.from_formula(formula, data=df,drop_absorbed=True)
res_fe=mod_fe.fit(cov_type='clustered', cluster_entity=True)
print(res_fe.summary.tables[1])
print(res_fe.summary.tables[0])
plot_result(res_fe,file_name="rx_log",title="RX log",period=25,insert_index=9,time_start=-10)

# formula="otc_dom ~ year_trend+elapsed_m10+elapsed_m9+elapsed_m8+elapsed_m7+elapsed_m6+elapsed_m5+elapsed_m4+elapsed_m3+elapsed_m2+elapsed_0+elapsed_1+elapsed_2+elapsed_3+elapsed_4+elapsed_5+elapsed_6+elapsed_7+elapsed_8+elapsed_9+elapsed_10+elapsed_11+elapsed_12+elapsed_13+elapsed_14+elapsed_15+generic_per+TimeEffects+EntityEffects"
# # stock lag version
# # formula+="+stock_otc_lag"
# # gemeric_per
# formula+=" + generic_per"
# mod_fe=PanelOLS.from_formula(formula, data=df, drop_absorbed=True)
# # mod_fe=PanelOLS.from_formula(formula, data=df, drop_absorbed=False)
# res_fe=mod_fe.fit(cov_type='clustered', cluster_entity=True)
# print(res_fe.summary.tables[1])
# print(res_fe.summary.tables[0])
# plot_result(res_fe,file_name="otc",title="OTC",period=25,insert_index=9,time_start=-10)
# take log
formula="log_otc_dom ~elapsed_m10+elapsed_m9+elapsed_m8+elapsed_m7+elapsed_m6+ elapsed_m5+elapsed_m4+elapsed_m3+elapsed_m2+elapsed_0+elapsed_1+elapsed_2+elapsed_3+elapsed_4+elapsed_5+elapsed_6+elapsed_7+elapsed_8+elapsed_9+elapsed_10+elapsed_11+elapsed_12+elapsed_13+elapsed_14+elapsed_15+generic_per+TimeEffects+EntityEffects"

# till 15
formula="log_otc_dom ~elapsed_m10+elapsed_m9+elapsed_m8+elapsed_m7+elapsed_m6+elapsed_m5+elapsed_m4+elapsed_m3+elapsed_m2+elapsed_0+elapsed_1+elapsed_2+elapsed_3+elapsed_4+elapsed_5+elapsed_6+elapsed_7+elapsed_8+elapsed_9+elapsed_10+elapsed_11+elapsed_12+elapsed_13+elapsed_14+elapsed_15+TimeEffects+EntityEffects"

# stock lag version
formula+="+log_stock_otc_lag"

# year trend
# formula+="+year_trend"

# gemeric_per
formula+=" + generic_per"

# generic_share_r
formula+=" + generic_share_r"
# generic_share_q
formula+=" + generic_share_q"

mod_fe=PanelOLS.from_formula(formula, data=df,drop_absorbed=True)
res_fe=mod_fe.fit(cov_type='clustered', cluster_entity=True)
# res_fe=mod_fe.fit()
# print(res_fe.summary.tables[1])
# print(res_fe.summary.tables[0])
print(res_fe)
plot_result(res_fe,file_name="otc_log",title="OTC log",period=25,insert_index=9,time_start=-10)

File: estimate/event/estimate_panel.r
# 必要なライブラリを読み込む
library(plm)
library(lmtest)
library(sandwich)
library(pretrends)

# データの読み込み
df <- read.csv("./data/merged/len3_ship_DN.csv", header = TRUE, encoding = "shift-jis")

# 必要な列を抽出・変換
df <- df[df$otc_dom > 0 & df$rx_dom > 0, ]
df$log_rx_dom <- log(df$rx_dom)
df$log_otc_dom <- log(df$otc_dom)
df$log_stock_rx_lag <- log(df$stock_rx_lag + 0.001)
df$log_stock_otc_lag <- log(df$stock_otc_lag + 0.001)
df$year_trend <- df$year - 2008

# パネルデータに変換
df_panel <- pdata.frame(df, index = c("id", "year"))

# モデル1の推定
formula1 <- log_rx_dom ~ year_trend + elapsed_m10 + elapsed_m9 + elapsed_m8 + elapsed_m7 + elapsed_m6 +
elapsed_m5 + elapsed_m4 + elapsed_m3 + elapsed_m2 + elapsed_0 + elapsed_1 + elapsed_2 + elapsed_3 +
elapsed_4 + elapsed_5 + elapsed_6 + elapsed_7 + elapsed_8 + elapsed_9 + elapsed_10 + elapsed_11 +
elapsed_12 + elapsed_13 + elapsed_14 + elapsed_15 + generic_per+log_stock_rx_lag + generic_share_r + generic_share_q

model1 <- plm(formula1, data = df_panel, model = "within", effect = "twoways")
summary_model1 <- coeftest(model1, vcov. = vcovHC(model1, type = "HC3"))
print(summary_model1)

# モデル2の推定
formula2 <- log_otc_dom ~ elapsed_m10 + elapsed_m9 + elapsed_m8 + elapsed_m7 + elapsed_m6 + elapsed_m5 +
elapsed_m4 + elapsed_m3 + elapsed_m2 + elapsed_0 + elapsed_1 + elapsed_2 + elapsed_3 + elapsed_4 +
elapsed_5 + elapsed_6 + elapsed_7 + elapsed_8 + elapsed_9 + elapsed_10 + elapsed_11 + elapsed_12 +
elapsed_13 + elapsed_14 + elapsed_15 + generic_per + log_stock_otc_lag +
generic_share_r + generic_share_q

model2 <- plm(formula2, data = df_panel, model = "within", effect = "twoways")
summary_model2 <- coeftest(model2, vcov. = vcovHC(model2, type = "HC3"))
print(summary_model2)
beta=summary_model2[1:25, 1]
sigma=vcovHC(model2, type = "HC3")[1:25, 1:25]
# -10 to 15 and drop the reference period
tVec=seq(-10,15,1)[-10]
referencePeriod <- -1 #This is the omitted period in the regression
data.frame(t = tVec, beta = beta)
slope50 <-
slope_for_power(sigma = sigma,
                targetPower = 0.8,
                tVec = tVec,
                referencePeriod = referencePeriod)
slope50

pretrendsResults <- 
  pretrends(betahat = beta, 
            sigma = sigma, 
            tVec = tVec, 
            referencePeriod = referencePeriod,
            deltatrue = slope50 * (tVec - referencePeriod))
pretrendsResults$event_plot
pretrendsResults$df_power
